api/v1alpha1/securitygroup_types.go

```diff
@@ -26,40 +26,88 @@ type RuleDirection string
+	ProtocolIPV6ENCAP Protocol = "ipv6-encap"
+	ProtocolIPV6FRAG  Protocol = "ipv6-frag"
+	ProtocolIPV6ICMP  Protocol = "ipv6-icmp"
+	ProtocolIPV6NONXT Protocol = "ipv6-nonxt"
+	ProtocolIPV6OPTS  Protocol = "ipv6-opts"
+	ProtocolIPV6ROUTE Protocol = "ipv6-route"
+	ProtocolOSPF      Protocol = "ospf"
+	ProtocolPGM       Protocol = "pgm"
+	ProtocolRSVP      Protocol = "rsvp"
+	ProtocolSCTP      Protocol = "sctp"
+	ProtocolTCP       Protocol = "tcp"
+	ProtocolUDP       Protocol = "udp"
+	ProtocolUDPLITE   Protocol = "udplite"
+	ProtocolVRRP      Protocol = "vrrp"
+)
+
 // +kubebuilder:validation:Enum:=IPv4;IPv6
 // +kubebuilder:validation:MinLength:=1
-// +kubebuilder:validation:MaxLength:=16
+// +kubebuilder:validation:MaxLength:=4
 type Ethertype string
 
+const (
+	EtherTypeIPv4 Ethertype = "IPv4"
+	EtherTypeIPv6 Ethertype = "IPv6"
+)
+
 // SecurityGroupRule defines a Security Group rule
 // +kubebuilder:validation:MinProperties:=1
+// +kubebuilder:validation:XValidation:rule="has(self.portRangeMin) == has(self.portRangeMax)",message="PortRangeMin and PortRangeMax should exist together"
+// +kubebuilder:validation:XValidation:rule="(!has(self.portRangeMin) || !has(self.portRangeMax)) || has(self.protocol)",message="Protocol must exist when PortRangeMin or PortRangeMax exist"
+// +kubebuilder:validation:XValidation:rule="(!has(self.portRangeMin) || !has(self.portRangeMax))|| (self.portRangeMin <= self.portRangeMax)",message="PortRangeMax should be equal or greater than PortRangeMin"
+// +kubebuilder:validation:XValidation:rule="!has(self.protocol) || !(self.protocol == 'icmp' || self.protocol == 'icmpv6') || (self.portRangeMin >= 0 && self.portRangeMin <= 255 && self.portRangeMax >= 0 && self.portRangeMax <= 255)",message="When protocol is ICMP or ICMPv6 PortRangeMin and PortRangeMax should be between 0 and 255"
+// +kubebuilder:validation:XValidation:rule="!has(self.remoteIPPrefix) || (isCIDR(self.remoteIPPrefix) && cidr(self.remoteIPPrefix).ip().family() == 4 && self.ethertype == 'IPv4') || (isCIDR(self.remoteIPPrefix) && cidr(self.remoteIPPrefix).ip().family() == 6 && self.ethertype == 'IPv6')",message="RemoteIPPrefix should be a valid CIDR and match the Ethertype"
 type SecurityGroupRule struct {
 	// Description of the existing resource
 	// +optional
+
 	Description *OpenStackDescription `json:"description,omitempty"`
 
 	// Direction represents the direction in which the security group rule
 	// is applied. Can be ingress or egress.
+	// +optional
 	Direction *RuleDirection `json:"direction,omitempty"`
 
 	// RemoteAddressGroupId (Not in gophercloud)
-
 	// RemoteGroupID
 	RemoteGroupID *UUID `json:"remoteGroupID,omitempty"`
 
-	// RemoteIPPrefix
+	// RemoteIPPrefix is an IP address block. Should match the Ethertype (IPv4 or IPv6)
+	// +optional
 	RemoteIPPrefix *CIDR `json:"remoteIPPrefix,omitempty"`
 
 	// Protocol is the IP protocol can be represented by a string, an
 	// integer, or null
+	// +optional
```

Comment:
(I'm commenting the comment) I don't believe you can represent the protocol as an integer here. You probably could if `type Protocol`'s underlying type was an integer and you overrode `UnmarshalJSON` to also accept strings.

Reply:
I'm not sure what the concern is. If you want vrrp to be set as a number you set the number "112" (as a string)

Reply:
If you want users to pass integers as strings, you should tell them explicitly in the comment. Also would null be passed as `"null"`?

Reply:
Regarding null I think I'll just remove it. About numbers as strings - Do we want to use also allow integers?

Reply:
that's a question for @mdbooth @mandre

Reply:
ok we'll go with not specified and use strings. As for enum - we use a pattern instead

Reply:
@mdbooth 
May I humour you with [this](https://bugs.launchpad.net/neutron/+bug/2074056/comments/1)

(also cf https://github.com/gophercloud/gophercloud/issues/2442#issuecomment-2249862493 )